### ISSUE TITLE
Revisit saga data locking

### ISSUE THREAD

**Issue by mookid8000** [rebus-org/Rebus#945]
See if the current locking abstraction is sufficient to enable plugging in a distributed locking mechanism of some kind.

Also look at whether correlation properties should/should not be included in the locking scheme.

**Comment by mookid8000** [comment]
The relevant interface is [`IExclusiveSagaAccessLock`](https://github.com/rebus-org/Rebus/blob/master/Rebus/Sagas/Exclusive/IExclusiveSagaAccessLock.cs) and it's good enough to manage distributed locks. 🙂

**Comment by mario-d-s** [comment]
@mookid8000 can you give an example of how consumers of Rebus are supposed to use it? Cause with the simple configuration line using the MySQL storage it did not work at all...

**Comment by mookid8000** [comment]
Check this out: Here's an example of how a custom lock implementation could look: https://github.com/rebus-org/Rebus/blob/master/Rebus.Tests/Sagas/TestSagaInstanceLocking_Custom.cs#L151-L166 (in this case simply using an ordinary semaphore as a mutex), and here's how it's used: https://github.com/rebus-org/Rebus/blob/master/Rebus.Tests/Sagas/TestSagaInstanceLocking_Custom.cs#L54

**Comment by mario-d-s** [comment]
@dietermijle @VDBBjorn this is related to that saga contention issue we had - I get the idea we should try to implement a custom exclusive lock that is based on the MySQL saga data table into the Rebus.MySql package. That would seem easy enough to do by leveraging transactions and acquiring a row lock or something.

@mookid8000 does that line of thinking make sense? I notice that you are also the maintainer of the Rebus.MySql package and I have briefly looked if it didn't already contain such lock implementation but none that I can find. If my reasoning is correct would you accept a PR for it? I think we will have time in a couple of weeks to look into it.

**Comment by mookid8000** [comment]
> (...) would you accept a PR for it?

Yes!! And yes, while I am officially the maintainer of all the packages under [rebus-org](https://github.com/rebus-org), I am not actually the one doing the work 😁 as I am not using MySQL myself, I'm lucky to have the super-productive @kendallb crunching away on the MySQL stuff 🤠 

@kendallb Do you happen to know anything about MySQL locking? I mean, I know you do, because that's essentially what the lease-based transport does 🙂

**Comment by kendallb** [comment]
I was looking at this not long ago, but we are not using Saga's yet so have not spent much time on it. It should be possible to set up a database lock with MySQL such that threads across multiple nodes can share in the lock. We do that all the time in our own code with a Mutex class we build that simply stores the lock status in a MySQL table, and then we use transactions to make sure only one thread at a time is able to get the lock. I expect you can do something similar with a lock column for the saga data and set the column to the ID of the caller for the duration of when they have the lock.

I implemented a similar thing just recently for the MySQL transport to support guaranteed ordering of messages coming out of MySQL across multiple nodes, by using an ordering key column. Only one thread at a time is able to write to the ordering key (the rest all stall out on the select), and once one thread has that message, any other messages in the queue simply get ignored until the first one is processed. That way we can still have multiple threads processing messages, but make sure messages for say a specific order only ever get processed in serial.

When I have a chance later in the week I could certainly take a look, but I expect it should be possible to implement this for our of process locks.

**Comment by kendallb** [comment]
The biggest problem I see with this interface at the moment, is the key that is passed in is constructed from the customer locking type and some other data, so the key in the case of the sample you wrote is this:

{ SagaDataType = Rebus.Tests.Sagas.TestSagaInstanceLocking_Custom+TypicalContendedSagaData, CorrelationPropertyName = Id, CorrelationPropertyValue = 1559011d-212a-4a9f-bafe-4436817da155 }

The catch with trying to do this with MySQL, is that in order to make it clean we need a database table row we can select on with a FOR UPDATE lock, so someone can set a 'locked' flag to indicate they have the lock, and others get locked out. My idea for this would be that the key passed in to the lock would actually be the GUID for the saga in question that is being locked, so we could then have a field in the MySQL Saga table that could be used to enforce the locking?

If it's a key as above, while you could have a table to store active locks in, it's much harder to get good concurrency without catching duplicate entry exceptions if the MySQL table is a separate table just for locks.

So I think for implementing this in MySQL, we would want a way to be able to export an interface from the MySQL Saga persistence layer itself that has the lock functions on it, that take the Saga ID as input. It would not be that easy to do it using MySQL with the current interface (could be done, but you then have multiple threads trying to insert and have to capture duplicate entry exceptions to know that someone else already got the lock).

Implementing a locking mechanism is something else like a key value store (MongoDB, Couchbase or something) is also another option, which would work with the existing interface where the key is generated from the data.

**Comment by kendallb** [comment]
The code to generate the lock ID key for the saga is as follows:

```C#
var locksToObtain = correlationProperties
    .Select(a => new
    {
        SagaDataType = a.saga.GetSagaDataType().FullName,
        CorrelationPropertyName = a.correlationProperty.PropertyName,
        CorrelationPropertyValue = a.correlationProperty.ValueFromMessage(messageContext, messageBody)
    })
    .Select(a => a.ToString())
    .OrderBy(str => str) // enforce consistent ordering to avoid deadlocks
    .ToArray();
```

Seems to me to be way overkill for this task, as the CorrelationPropertyValue that comes out of this is the Saga ID GUID, so it's already unique? Or there some other reason the CorrelationPropertyValue would not be the GUID of the Saga in question? If we can rely on it being the GUID for the Saga in question, then you could do locking on the Saga row. But I assume it was written this way for a reason, so perhaps that value may not actually be the ID of the Saga? 

Would it be possible to always get the GUID ID of the Saga being locked to pass to the IExclusiveSagaAccessLock as that would make it easier to implement the locking.

Although in my initial testing using the MySQL Saga layer, it looks like the first time the lock is requested the Saga is not actually persisted anywhere so the first thread in gets the lock, and then the Saga is created and the others all block until they get their chance to get in. So the concept of somehow doing a row lock (or a lock flag in the row) won't quite work. But perhaps we could do something like this:

1. Check for the presence of the Saga in the table by GUID
2. If the entry is not present, lock is 'aquired'
2. When new Saga entries are inserted, the lock column is always '1'
3. When checking for a lock when the Saga entry already exists, if the lock is '1', block
4. When the lock is released, set the lock value for the Saga entry to '0'
5. Next thread in the locking code gets in and sets the value to '1' and aquires the lock
6. Repeat until done

The caveat with all this is that there is no way to do an async block on the 'aquire' code that I can think of easily with MySQL? The only way an actual row lock using FOR UPDATE would work is if a MySQL transaction is held open until the Saga lock is released, and then I guess you could try doing an update which would block or time out.

Otherwise the lock could would probably need to do some kind of loop with a short sleep?

**Comment by kendallb** [comment]
Ok something is not right here when I test this. The first AquireLockAsync() call that comes in has a GUID for the correlation property value that is NOT the ID of the Saga that ends up in the table. When the ReleaseLockAsync() call comes in, it is using the key that was originally used for the lock, but subsequent AquireLockAsync() calls DO have the Saga ID as the correlation property value?

Although the first AquireLockAsync() and ReleaseLockAsync() calls come in before all the others, so perhaps when the Saga is being initially created we could just check for the Saga and assume it's all good because it does not exist in the table, and then all the other ones will be using the same lock? 

Certainly the way it is at the moment the first lock is using a totally different key to all the subsequent ones, but perhaps that is expected if the Saga was not yet started?

**Comment by mookid8000** [comment]
To avoid contention around the creation of new saga instances, locking is NOT done on the saga's ID, it is done on correlation properties. Lock IDs are created out of the saga's type, a saga data correlation property name, and the corresponding value from the incoming message.

This e.g. means that two different messages types, both looking to be correlated via an "order number property" could result in competing for a lock named `{ SagaDataType = MyCode.ContendedSagaData, CorrelationPropertyName = OrderNumber, CorrelationPropertyValue = 3746d2d4-c2cb-4f9f-a95b-f827edf5286d }` that is structured like this:

```
//                  \/-saga type                                      \/- correlation property name                     \/- value from message
{ SagaDataType = MyCode.ContendedSagaData, CorrelationPropertyName = OrderNumber, CorrelationPropertyValue = 3746d2d4-c2cb-4f9f-a95b-f827edf5286d }
```

Does that make sense?

**Comment by kendallb** [comment]
Yep, makes sense. A separate lock table will work then especially since the first lock in has to create the entry. Just have to capture duplicate key exceptions and try again when that happens as it means the other thread got In before you.

**Comment by kendallb** [comment]
Working an implementation for this. What would be the best way to hook my new class into the injectionist, so that it can export IInitializable and IDisposable? I want it to auto create the necessary tables as well as have a regular cleanup service.

**Comment by kendallb** [comment]
Ok worked it out. I need some changes to Rebus to make this work, as we need a way to be able to register IExclusiveSagaAccessLock with the injectionist and then use that if it's found in EnforceExclusiveAccess(). Currently it either will create the internal one using SemaphoreSlim, or it will require you to construct your own instance and pass that in. But there is no way to set up a class to be registered with the injectionist so it's lifecycle can be managed.

I have a patch for that I can put into a PR.

**Comment by kendallb** [comment]
https://github.com/rebus-org/Rebus/pull/960

**Comment by kendallb** [comment]
Inside EnforceExclusiveSagaAccessIncomingStep the code is like this:

```C#
async Task WaitForLocks(string[] lockIds)
{
    for (var index = 0; index < lockIds.Length; index++)
    {
        while (!await _lockHandler.AquireLockAsync(lockIds[index], _cancellationToken))
        {
            await Task.Yield();
        }
    }
}

async Task ReleaseLocks(string[] lockIds)
{
    for (var index = 0; index < lockIds.Length; index++)
    {
        await _lockHandler.ReleaseLockAsync(lockIds[index]);
    }
}
```

That is probably not a good idea in the case of a lock stores in something like MySQL, because if the lock is not aquired it will loop constantly and all it does is yield the current thread. That will hammer the crap out of the SQL data store until the lock gets aquired, so we probably want to actually have some kind of thread sleep there? 

If I compare it to how the 'NewEnforceExclusiveSagaAccessIncomingStep' is implemented (the default unless you create an instance of a locker class), that blocks the thread until the semaphore is released. Since there is no easy way that I can think of to simulate a blocking operation with a MySQL read command, I think the only viable solution is an async thread sleep. I could implement that in the MySQL locker, but I think it should probably be done in the generic code in Rebus? 

Unless you don't like it, I will add it to my PR above.

```C#
async Task WaitForLocks(List<int> lockIds)
{
    for (var index = 0; index < lockIds.Count; index++)
    {
        var id = lockIds[index];
        await _locks[id].WaitAsync(_cancellationToken);
    }
}

void ReleaseLocks(List<int> lockIds)
{
    for (var index = 0; index < lockIds.Count; index++)
    {
        var id = lockIds[index];
        _locks[id].Release();
    }
}
```

**Comment by kendallb** [comment]
Ok code is written and it works nicely. Due to the fact that EnforceExclusiveSagaAccessIncomingStep is not publicly accessible, and the internals required to implement it are also not public, this cannot work until some changes are made in Rebus itself.

**Comment by kendallb** [comment]
Ahh, looks like I can use GET_LOCK for MySQL to avoid all the thread delays. I have not looked at using that in some time, but last time was prior to MySQL 5.7 and it was pretty useless (only one lock at a time), but it looks like that has been changed since 5.7:

https://dev.mysql.com/doc/refman/5.7/en/locking-functions.html#function_get-lock

So should be able to change the implementation here to use GET_LOCK and have the MySQL connection do the blocking async.

**Comment by kendallb** [comment]
Gah, not sure this will work. Not safe for statement based replication and other caveats... MySQL just does not seem to have the equivalent of sp_getapplock from SQL Server.

**Comment by mookid8000** [comment]
> That is probably not a good idea in the case of a lock stores in something like MySQL, because if the lock is not aquired it will loop constantly and all it does is yield the current thread. That will hammer the crap out of the SQL data store until the lock gets aquired, so we probably want to actually have some kind of thread sleep there?

As you already seem to have figured out, it is totally fine if the lock implementation does not return from the call to `WaitAsync` immediately – it just needs to respect the `CancellationToken` passed to it.


> So the super long lock names won't work, so we would need to somehow hash them to a unique 64-character lock? Given how the lock keys are created (it is not always a GUID), what are the options here?

If the lock implementation has a contraint on the length of the lock names (or other constraints, e.g. around use of invalid characters etc.) then it must simply apply some kind of deterministic transformation on the lock ID to make it valid.

The important thing here is not that the lock IDs are unique – in fact, I can imagine many cases where it would not make sense to use a 1-1 relationship between the lock ID and the actually used ID, because it would leave some kind of entity back in whatever mechanism is used to implement the locking.

For example, because the lock ID space is infinite, one could get a Knuth hash of the lock ID and do a modulo on it, mapping it into a known finite space, e.g. like so:

```
lock ID space     \
                   \
...                 \
                     \
...                     ____________________ 0
               

....                                         ...
               
               			actual lock ID space
                             
                                             ...
...               
               
                        ____________________ 1023
...               
                     /   
                    /
                   /
...               /
```

which would remove all contention around saga instances, reducing it to a potential contention around specific locks. In this case, there's 1024 locks, and if the transformation function is uniformly distributed, then there's a good chance that lots of parallel stuff can still happen without any collisions.

**Comment by kendallb** [comment]
Ahh good idea. I need an infinite space for something else I am working on (basically async replies across distributed nodes), but a finite lock space is similar to what you did on the new semaphore slim version. I will make some changes to make it work that way as it does make sense that if there is a clash, its OK if some unrelated sagas get blocked. All you really need is more locks than message threads so 1024 is probably more than enough. 

I don’t think it can work though without some changes to rebus itself but I can remove the delay stuff from there and implement it in the locker itself (and probably put some random jitter in there as well).

Not sure we can use GET_LOCK though due to the limitations on it working with slaving. I think most folks use bin log replication anyway so perhaps there is a way to check if statement based replication is enabled and throw an error in that case (and also check for version 5.7 and later).

**Comment by kendallb** [comment]
Ok so the other problem with GET_LOCK is that it only persists for the client session, which means it only persists for the duration of the connection. If we create a connection and throw it away, it won't work. So not sure this is really going to work and will have to stick to the table mechanism I have already created.

As for a limited lock space, I thin it is still a good idea to do that but it cannot be done at the locker level because of the potential to end up trying to get the same lock twice in the EnforceExclusiveSagaAccessIncomingStep. So I think the solution there is to change the IExclusiveSagaAccessLock locking API to not take a string lock key at all, but to take an integer lock key and build the implementation similiar to the SemaphoreSlim version? Can even reuse the class across both. So I think I will re-factor the code changes I made in Rebus to do that.

**Comment by kendallb** [comment]
Ok I have modified the pull request for Rebus itself as outlined above:

https://github.com/rebus-org/Rebus/pull/960

I changed the lock keys to integers and use a hashed lock ID range for all locks similar to the sempahore slim version, as it is not possible to do that in the locker itself (due to the potential for hash clashes and trying to get the same lock twice). Also that PR adds a way for the exclusive saga lock instance to be pulled from injectionist so we can register it and have it's lifecycle managed. It uses that to make sure the SQL tables are created as necessary and that a backgrounds thread is designed to run every 5 minutes to clean up any dangling locks.

The locker step implementation was changed to first check if it can find the interface via the injectionist and use that if it is there,  otherwise it will fall back on the built in semaphore slim method if not.

Not sure if the breaking change to change the lock key to a string is a problem, but if it is desired to leave it as a string then the interface could go back to a string and we would do a .ToString() on the integer lock ID before it's passed to the locker.

The PR for the exclusive locks is here:

https://github.com/rebus-org/Rebus.MySql/pull/19

But it will not build without the changes from the PR above being added to Rebus itself. I also expose a separate interface for locking that does not do the thread sleeps, uses string lock keys and also exposes a separate interface that has a function to check if the lock has been aquired or not and not actually aquire it.

I will be using that interface in our own code to implement some changes to the Rebus.Async library, to use that rather than a concurrent dictionary to signal to blocks threads waiting on a reply to wake up. That will then allow that library to work across distributed nodes, not just on one node.

**Comment by kendallb** [comment]
Ok changed direction on how to set up the locking (after I tried to use it for Async) so that now IExclusiveAccessLock is now an interface exposed by Rebus and has a configuration section to control it. If it is not registered, the saga locks use in process semaphore slims to do the locking, but you can choose to use the older concurrent dictionary model now (which is there mostly to test everything and make sure it works). 

New PR is here and I killed the old one off (code is still in Git though for now):

https://github.com/rebus-org/Rebus/pull/962

I will work on the MySQL version of it next.

**Comment by kendallb** [comment]
Ok, PR has been updated for MySQL to use the new functionality. At this point I think it is close to done :)

https://github.com/rebus-org/Rebus.MySql/pull/19

**Comment by kendallb** [comment]
I need access to IExclusiveAccessLock for my own code (the entire reason I wrote it in the first place :) ), but I could not find a clean way to get access to things resolved internally so I changed Rebus to expose it via the advanced bus API, as that seemed like the correct way to expose it as a public API. If there is a better way, please let me know. For now this gets me moving ahead with using it for UI synchronization messages which is what I need it for (to ensure all messages for an ordering key are processed up to that point).

Now ConcurrentDictionaryExclusiveAccessLock is the default locker implementation and you have to specifically ask to use the locker for Sagas, otherwise the default is to use the semaphore slim model. You would really only change to use the locker implementation if you set the locker up as something else, like MySQL since it is not as good as semaphores, but those are only useful for in process locking. Code in the PR is updated.

**Comment by kendallb** [comment]
Ok, got it all working nicely in our code now. I will probably publish my own packages until you can get around to reviewing the code and deciding if you like my approach or want to change it. Will likely push this into production on our end on Monday.

**Comment by kendallb** [comment]
What is the ETA on getting Rebus 7.0.0 out, and pushing Rebus.MySql 3.0.0 as official releases?

**Comment by mookid8000** [comment]
In a couple of days, if 7.0.0-b02 works as it should 🙂
